tests/v2/validation/provisioning/rke2/etcd_bkp_restore_test.go

```diff
@@ -0,0 +1,373 @@
+package rke2
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/rancher/norman/types"
+	"github.com/rancher/rancher/tests/framework/clients/rancher"
+	provisioningV1 "github.com/rancher/rancher/tests/framework/clients/rancher/v1"
+	"github.com/rancher/rancher/tests/framework/extensions/clusters"
+	"github.com/rancher/rancher/tests/framework/extensions/ingresses"
+	"github.com/rancher/rancher/tests/framework/extensions/machinepools"
+	"github.com/rancher/rancher/tests/framework/extensions/workloads"
+	"github.com/rancher/rancher/tests/framework/extensions/workloads/pods"
+	"github.com/rancher/rancher/tests/integration/pkg/defaults"
+
+	"github.com/rancher/rancher/tests/framework/pkg/config"
+	"github.com/rancher/rancher/tests/framework/pkg/session"
+	"github.com/rancher/rancher/tests/framework/pkg/wait"
+	provisioning "github.com/rancher/rancher/tests/v2/validation/provisioning"
+
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	kubeProvisioning "github.com/rancher/rancher/tests/framework/clients/provisioning"
+	appv1 "k8s.io/api/apps/v1"
+	networkingv1 "k8s.io/api/networking/v1"
+	kwait "k8s.io/apimachinery/pkg/util/wait"
+)
+
+const (
+	defaultNamespace = "default"
+)
+
+type RKE2EtcdSnapshotRestoreTestSuite struct {
+	suite.Suite
+	session            *session.Session
+	client             *rancher.Client
+	ns                 string
+	kubernetesVersions []string
+	cnis               []string
+	providers          []string
+	nodesAndRoles      []machinepools.NodeRoles
+}
+
+func (p *RKE2EtcdSnapshotRestoreTestSuite) TearDownSuite() {
+	p.session.Cleanup()
+}
+
+func (r *RKE2EtcdSnapshotRestoreTestSuite) SetupSuite() {
+	testSession := session.NewSession(r.T())
+	r.session = testSession
+
+	r.ns = defaultNamespace
+
+	clustersConfig := new(provisioning.Config)
+	config.LoadConfig(provisioning.ConfigurationFileKey, clustersConfig)
+
+	r.kubernetesVersions = clustersConfig.KubernetesVersions
+	r.cnis = clustersConfig.CNIs
+	r.providers = clustersConfig.Providers
+	r.nodesAndRoles = clustersConfig.NodesAndRoles
+
+	client, err := rancher.NewClient("", testSession)
+	require.NoError(r.T(), err)
+
+	r.client = client
+
+}
+
+func (r *RKE2EtcdSnapshotRestoreTestSuite) upgradeClusterK8sVersion(client *rancher.Client, clustername string, k8sUpgradedVersion string) error {
```

Comment:
agree

Reply:
In my opinion, this can't be moved to the `etcd_backup_restore.go` as it is. This is a private method that uses the suite struct. In go, methods are usually contained in the same place as the structs. If the action/method is to be separated, then they are separated by their naming or logic. To move this one, either method part needs to be changed as:

```
func upgradeClusterK8sVersion(t *testing.T, client *rancher.Client, clusterName string, namespaceName string,  k8sUpgradedVersion string) error { 
t.Helper()
// get namespace via passing as a function argument and use testing T by passing current testing T, use testing T dependent assertions
}
```

or 

```
func upgradeClusterK8sVersion(client *rancher.Client, clusterName string, namespaceName string, k8sUpgradedVersion string) error { 
// get namespace via passing as a function argument and change testing T dependent things to error/catch and so forth
}
```

Reply:
That's an easy change though, just have the namespace be an parameter. I do think it needs to be move with a slight modification

Reply:
Suite-related things aren't in use that much, I agree with moving those with a few changes. I have the same opinion about the rest of the discussed methods.

Reply:
Made the required changes.